File: api/routes/sessions.py
import uuid
import datetime
import os
import requests
from flask import Blueprint, request, jsonify, g
from ..db.database import get_db

sessions_bp = Blueprint("sessions", __name__, url_prefix="/api")

VLLM_BASE_URL = os.getenv("VLLM_BASE_URL", "http://vllm.internal/v1")
OPENAI_API_KEY = os.getenv("OPENAI_API_KEY", "")

# OpenAI 모델명 → 사외 테스트용 모델 식별
OPENAI_MODELS = {"gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "gpt-3.5-turbo"}


def row_to_dict(row):
    return dict(row) if row else None


def rows_to_list(rows):
    return [dict(row) for row in rows]


def _chat_completions(base_url: str, api_key: str, model: str, messages: list, max_tokens: int = 30) -> str:
    headers = {"Content-Type": "application/json"}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"
    resp = requests.post(
        f"{base_url}/chat/completions",
        headers=headers,
        json={"model": model, "messages": messages, "max_tokens": max_tokens, "temperature": 0.3},
        timeout=10,
    )
    resp.raise_for_status()
    return resp.json()["choices"][0]["message"]["content"].strip()


def generate_title(message: str, model: str) -> str:
    messages = [
        {"role": "system", "content": "다음 메시지를 보고 5단어 이내 한국어 채팅 제목을 만들어라. 제목만 출력하라."},
        {"role": "user", "content": message[:500]},
    ]
    # OpenAI 모델이거나 OPENAI_API_KEY가 있으면 OpenAI 사용
    if model in OPENAI_MODELS or (OPENAI_API_KEY and VLLM_BASE_URL == "http://vllm.internal/v1"):
        title_model = model if model in OPENAI_MODELS else "gpt-4o-mini"
        try:
            return _chat_completions("https://api.openai.com/v1", OPENAI_API_KEY, title_model, messages)
        except Exception:
            pass
    # 사내 vLLM 시도
    try:
        return _chat_completions(VLLM_BASE_URL, "", model, messages)
    except Exception:
        return message[:30]


@sessions_bp.get("/sessions")
def list_sessions():
    """최근 순 세션 목록 반환."""
    try:
        db = get_db()
        rows = db.execute(
            """
            SELECT id, title, persona, model, created_at, updated_at
            FROM sessions
            WHERE user_id = ?
            ORDER BY updated_at DESC
            LIMIT 100
            """,
            (g.user_id,),
        ).fetchall()
        return jsonify(rows_to_list(rows))
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.post("/sessions")
def create_session():
    """새 세션 생성."""
    try:
        body = request.get_json(silent=True) or {}
        session_id = str(uuid.uuid4())
        now = datetime.datetime.utcnow().isoformat()
        title = body.get("title", "새 대화")
        persona = body.get("persona")
        model = body.get("model", "GLM4.7")

        db = get_db()
        db.execute(
            """
            INSERT INTO sessions (id, user_id, title, persona, model, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?)
            """,
            (session_id, g.user_id, title, persona, model, now, now),
        )
        db.commit()

        row = db.execute(
            "SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE id = ?",
            (session_id,),
        ).fetchone()
        return jsonify(row_to_dict(row)), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.get("/sessions/<session_id>")
def get_session(session_id):
    """세션 정보 + 메시지 이력 반환."""
    try:
        db = get_db()
        session = db.execute(
            "SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE id = ?",
            (session_id,),
        ).fetchone()

        if session is None:
            return jsonify({"error": "Session not found"}), 404

        messages = db.execute(
            "SELECT id, role, content, created_at FROM messages WHERE session_id = ? ORDER BY created_at ASC",
            (session_id,),
        ).fetchall()

        result = row_to_dict(session)
        result["messages"] = rows_to_list(messages)
        return jsonify(result)
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.delete("/sessions/<session_id>")
def delete_session(session_id):
    """세션 삭제 (메시지 포함, CASCADE)."""
    try:
        db = get_db()
        row = db.execute("SELECT id FROM sessions WHERE id = ?", (session_id,)).fetchone()
        if row is None:
            return jsonify({"error": "Session not found"}), 404

        db.execute("DELETE FROM sessions WHERE id = ?", (session_id,))
        db.commit()
        return jsonify({"deleted": session_id})
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.patch("/sessions/<session_id>")
def update_session(session_id):
    """세션 title 업데이트."""
    try:
        body = request.get_json(silent=True) or {}
        db = get_db()

        row = db.execute("SELECT id FROM sessions WHERE id = ?", (session_id,)).fetchone()
        if row is None:
            return jsonify({"error": "Session not found"}), 404

        now = datetime.datetime.utcnow().isoformat()
        updates = []
        values = []

        if "title" in body:
            updates.append("title = ?")
            values.append(body["title"])
        if "persona" in body:
            updates.append("persona = ?")
            values.append(body["persona"])
        if "model" in body:
            updates.append("model = ?")
            values.append(body["model"])

        if not updates:
            return jsonify({"error": "No fields to update"}), 400

        updates.append("updated_at = ?")
        values.append(now)
        values.append(session_id)

        db.execute(
            f"UPDATE sessions SET {', '.join(updates)} WHERE id = ?",
            values,
        )
        db.commit()

        updated = db.execute(
            "SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE id = ?",
            (session_id,),
        ).fetchone()
        return jsonify(row_to_dict(updated))
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.post("/sessions/<session_id>/messages")
def append_message(session_id):
    """메시지 추가."""
    try:
        body = request.get_json(silent=True) or {}
        role = body.get("role")
        content = body.get("content")

        if not role or content is None:
            return jsonify({"error": "role and content are required"}), 400

        if role not in ("system", "user", "assistant", "tool"):
            return jsonify({"error": "Invalid role"}), 400

        db = get_db()
        session = db.execute("SELECT id FROM sessions WHERE id = ?", (session_id,)).fetchone()
        if session is None:
            return jsonify({"error": "Session not found"}), 404

        msg_id = str(uuid.uuid4())
        now = datetime.datetime.utcnow().isoformat()

        db.execute(
            "INSERT INTO messages (id, session_id, role, content, created_at) VALUES (?, ?, ?, ?, ?)",
            (msg_id, session_id, role, content, now),
        )
        db.execute(
            "UPDATE sessions SET updated_at = ? WHERE id = ?",
            (now, session_id),
        )
        db.commit()

        msg = db.execute(
            "SELECT id, session_id, role, content, created_at FROM messages WHERE id = ?",
            (msg_id,),
        ).fetchone()
        return jsonify(row_to_dict(msg)), 201
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.post("/sessions/generate-title")
def generate_title_endpoint():
    """LLM으로 5단어 이내 세션 제목 생성."""
    try:
        body = request.get_json(silent=True) or {}
        message = body.get("message", "")
        model = body.get("model", "GLM4.7")

        if not message:
            return jsonify({"error": "message is required"}), 400

        title = generate_title(message, model)
        return jsonify({"title": title})
    except Exception as e:
        return jsonify({"error": str(e)}), 500


@sessions_bp.post("/sessions/<session_id>/upload")
def upload_file(session_id):
    """파일 업로드 스텁 (Phase 8용)."""
    try:
        db = get_db()
        session = db.execute("SELECT id FROM sessions WHERE id = ?", (session_id,)).fetchone()
        if session is None:
            return jsonify({"error": "Session not found"}), 404

        if "file" not in request.files:
            return jsonify({"error": "No file provided"}), 400

        file = request.files["file"]
        filename = file.filename or "unnamed"

        # Phase 8: 실제 파일 파싱 및 임베딩 처리 예정
        # 현재는 스텁으로 파일명만 반환
        return jsonify({
            "status": "stub",
            "session_id": session_id,
            "filename": filename,
            "message": "File upload will be fully implemented in Phase 8",
        }), 202
    except Exception as e:
        return jsonify({"error": str(e)}), 500

File: server/routes/sessions.ts
import { Router } from "express";
import db from "../db.js";

const router = Router();
const VLLM_BASE_URL = process.env.VLLM_BASE_URL ?? "http://vllm.internal/v1";
const OPENAI_API_KEY = process.env.OPENAI_API_KEY ?? "";
const OPENAI_MODELS = new Set(["gpt-4o", "gpt-4o-mini", "gpt-4-turbo", "gpt-3.5-turbo"]);

function uid(req: import("express").Request): string {
  return (req.headers["x-user-id"] as string) ?? "default";
}

async function callLLM(baseUrl: string, apiKey: string, model: string, messages: object[]): Promise<string> {
  const headers: Record<string, string> = { "Content-Type": "application/json" };
  if (apiKey) headers["Authorization"] = `Bearer ${apiKey}`;
  const resp = await fetch(`${baseUrl}/chat/completions`, {
    method: "POST",
    headers,
    body: JSON.stringify({ model, messages, max_tokens: 30, temperature: 0.3 }),
    signal: AbortSignal.timeout(10000),
  });
  if (!resp.ok) throw new Error(`HTTP ${resp.status}`);
  const data = await resp.json() as { choices: Array<{ message: { content: string } }> };
  return data.choices[0].message.content.trim();
}

// POST /api/sessions/generate-title — 반드시 /:id 보다 먼저 등록
router.post("/sessions/generate-title", async (req, res) => {
  try {
    const { message = "", model = "GLM4.7" } = req.body ?? {};
    if (!message) return res.status(400).json({ error: "message is required" });
    const messages = [
      { role: "system", content: "다음 메시지를 보고 5단어 이내 한국어 채팅 제목을 만들어라. 제목만 출력하라." },
      { role: "user", content: (message as string).slice(0, 500) },
    ];
    let title = (message as string).slice(0, 30);
    try {
      if (OPENAI_MODELS.has(model) && OPENAI_API_KEY) {
        title = await callLLM("https://api.openai.com/v1", OPENAI_API_KEY, model, messages);
      } else {
        title = await callLLM(VLLM_BASE_URL, "", model, messages);
      }
    } catch { /* 실패 시 첫 30자 */ }
    res.json({ title });
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// GET /api/sessions
router.get("/sessions", (req, res) => {
  try {
    const rows = db.prepare(
      "SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE user_id = ? ORDER BY updated_at DESC LIMIT 100"
    ).all(uid(req));
    res.json(rows);
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// POST /api/sessions
router.post("/sessions", (req, res) => {
  try {
    const body = req.body ?? {};
    const id = crypto.randomUUID();
    const now = new Date().toISOString();
    db.prepare(
      "INSERT INTO sessions (id, user_id, title, persona, model, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?, ?)"
    ).run(id, uid(req), body.title ?? "새 대화", body.persona ?? null, body.model ?? "GLM4.7", now, now);
    res.status(201).json(
      db.prepare("SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE id = ?").get(id)
    );
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// GET /api/sessions/:id
router.get("/sessions/:id", (req, res) => {
  try {
    const session = db.prepare(
      "SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE id = ?"
    ).get(req.params.id);
    if (!session) return res.status(404).json({ error: "Session not found" });
    const messages = db.prepare(
      "SELECT id, role, content, created_at FROM messages WHERE session_id = ? ORDER BY created_at ASC"
    ).all(req.params.id);

    // tool_calls를 message_id 기준으로 그룹핑
    const toolCallRows = db.prepare(
      "SELECT id, message_id, tool_name, tool_label, args, result, is_error, started_at, ended_at, order_idx FROM tool_calls WHERE session_id = ? ORDER BY order_idx ASC"
    ).all(req.params.id) as Array<{
      id: string; message_id: string; tool_name: string; tool_label: string;
      args: string; result: string; is_error: number;
      started_at: string; ended_at: string; order_idx: number;
    }>;

    const toolCallsByMessage: Record<string, object[]> = {};
    for (const tc of toolCallRows) {
      if (!toolCallsByMessage[tc.message_id]) toolCallsByMessage[tc.message_id] = [];
      toolCallsByMessage[tc.message_id].push({
        id: tc.id,
        toolName: tc.tool_name,
        toolLabel: tc.tool_label,
        args: tc.args ? JSON.parse(tc.args) : {},
        result: tc.result ? JSON.parse(tc.result) : null,
        isError: tc.is_error === 1,
        startedAt: tc.started_at,
        endedAt: tc.ended_at,
        orderIdx: tc.order_idx,
      });
    }

    res.json({ ...(session as object), messages, toolCallsByMessage });
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// PATCH /api/sessions/:id
router.patch("/sessions/:id", (req, res) => {
  try {
    if (!db.prepare("SELECT id FROM sessions WHERE id = ?").get(req.params.id))
      return res.status(404).json({ error: "Session not found" });
    const body = req.body ?? {};
    const fields: string[] = [];
    const values: unknown[] = [];
    for (const f of ["title", "persona", "model"]) {
      if (f in body) { fields.push(`${f} = ?`); values.push(body[f]); }
    }
    if (!fields.length) return res.status(400).json({ error: "No fields to update" });
    fields.push("updated_at = ?");
    values.push(new Date().toISOString(), req.params.id);
    db.prepare(`UPDATE sessions SET ${fields.join(", ")} WHERE id = ?`).run(...values);
    res.json(db.prepare("SELECT id, title, persona, model, created_at, updated_at FROM sessions WHERE id = ?").get(req.params.id));
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// DELETE /api/sessions/:id
router.delete("/sessions/:id", (req, res) => {
  try {
    if (!db.prepare("SELECT id FROM sessions WHERE id = ?").get(req.params.id))
      return res.status(404).json({ error: "Session not found" });
    db.prepare("DELETE FROM sessions WHERE id = ?").run(req.params.id);
    res.json({ deleted: req.params.id });
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// POST /api/sessions/:id/messages
router.post("/sessions/:id/messages", (req, res) => {
  try {
    const body = req.body ?? {};
    if (!body.role || body.content === undefined)
      return res.status(400).json({ error: "role and content are required" });
    if (!["system", "user", "assistant", "tool"].includes(body.role))
      return res.status(400).json({ error: "Invalid role" });
    if (!db.prepare("SELECT id FROM sessions WHERE id = ?").get(req.params.id))
      return res.status(404).json({ error: "Session not found" });
    const msgId = crypto.randomUUID();
    const now = new Date().toISOString();
    db.prepare("INSERT INTO messages (id, session_id, role, content, created_at) VALUES (?, ?, ?, ?, ?)").run(
      msgId, req.params.id, body.role, body.content, now
    );
    db.prepare("UPDATE sessions SET updated_at = ? WHERE id = ?").run(now, req.params.id);
    res.status(201).json(
      db.prepare("SELECT id, session_id, role, content, created_at FROM messages WHERE id = ?").get(msgId)
    );
  } catch (err) {
    res.status(500).json({ error: String(err) });
  }
});

// POST /api/sessions/:id/upload (Phase 8 스텁)
router.post("/sessions/:id/upload", (req, res) => {
  if (!db.prepare("SELECT id FROM sessions WHERE id = ?").get(req.params.id))
    return res.status(404).json({ error: "Session not found" });
  res.status(202).json({ status: "stub", session_id: req.params.id, message: "File upload will be implemented in Phase 8" });
});

export default router;
